src/openstack_manager.py

```diff
@@ -0,0 +1,110 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for handling interactions with OpenStack."""
+import logging
+from pathlib import Path
+
+import keystoneauth1.exceptions
+import openstack
+import openstack.exceptions
+import yaml
+from openstack.identity.v3.project import Project
+
+from errors import OpenStackInvalidConfigError, OpenStackUnauthorizedError
+
+logger = logging.getLogger(__name__)
+
+CLOUDS_YAML_PATH = Path(Path.home() / ".config/openstack/clouds.yaml")
+
+
+def _validate_cloud_config(cloud_config: dict) -> None:
+    """Validate the format of the cloud configuration.
+
+    Args:
+        cloud_config: The configuration in clouds.yaml format to validate.
+
+    Raises:
+        InvalidConfigError: if the format of the config is invalid.
+    """
+    # dict of format: {clouds: <cloud-name>: <cloud-config>}
+    try:
+        clouds = list(cloud_config["clouds"].keys())
```

Comment:
All sorts of things could be in `clouds`, it doesn't necessarily have to be iterable which is what i think is required by the `list` constructor. Given that we know that it is going to come from a `yaml`, perhaps check it is a `list`? Also indicates that we should do additional testing here, e.g., passing invalid cloud config yaml in all ways that we can think of

Reply:
We're checking for the `keys()` of the dict, so I think as long as the validation of `dict` for cloud_config is done w/ reference to the comment above, it should be okay :D thank you for the review!